README.md
[![Hex Version](http://img.shields.io/hexpm/v/dinero.svg?style=flat)](https://hex.pm/packages/dinero)
[![Hex docs](http://img.shields.io/badge/hex.pm-docs-green.svg?style=flat)](https://hexdocs.pm/dinero)
[![Build Status](https://travis-ci.org/MLSDev/dinero.svg?branch=master)](https://travis-ci.org/MLSDev/dinero)

# Dinero

Elixir library for working with monetary values. Dinero is slang for money.

## How to use
```elixir
  iex> ten_bucks = Dinero.new(10, :USD)
  %Dinero{amount: 1000, currency: :USD}
  iex> twelve_bucks = Dinero.new(12, :USD)
  %Dinero{amount: 1200, currency: :USD}
  iex> two_bucks = Dinero.subtract(twenty_bucks, ten_bucks)
  %Dinero{amount: 200, currency: :USD}
  iex> four_bucks = Dinero.multiply(two_bucks, 2)
  %Dinero{amount: 400, currency: :USD}
  iex> convert = Dinero.convert(four_bucks, :UAH, 26.2)
  %Dinero{amount: 10480, currency: :UAH}
  iex> d = ~m[924.06]uah
  %Dinero{amount: 92406, currency: :UAH}
  iex> "Amount #{d}"
  "Amount 924.06"
  iex> Dinero.new(1.0e4, :USD) 
  %Dinero{amount: 1000000, currency: :USD}
  iex> Dinero.parse("1312.14")
  %Dinero{amount: 131214, currency: :USD}
```

### Using with Postgres
`Dinero.Ecto.DineroWithCurrency` is an Ecto type that allows saving amount of dinero with currency to database.

1. Create custom Postrgres type
    ```elixir
    def up do
      execute "CREATE TYPE public.dinero_with_currency AS (amount integer, currency char(3))"
    end
    
    def down do
      execute "DROP TYPE public.dinero_with_currency"
    end
    ```
2. Add column to the table
    ```elixir
    alter table (:products) do
      add :price, :dinero_with_currency
    end
    ```
3. Add field to the schema
    ```elixir
    schema "products" do
      field :price, Dinero.Ecto.DineroWithCurrency
    end
    ```
4. Save to DB:
    ```elixir
    iex> Repo.insert %Product{price: Dinero.new(20, :USD)}
    [debug] QUERY OK db=2.6ms decode=0.5ms queue=0.4ms
      INSERT INTO "products" ("price","inserted_at","updated_at") VALUES ($1,$2,$3) RETURNING "id" [{2000, "USD"}, ~N[2019-07-15 12:47:38], ~N[2019-07-15 12:47:38]]
    {:ok,
      %EctoCustomType.Product{
      __meta__: #Ecto.Schema.Metadata<:loaded, "products">,
      id: 1,
      inserted_at: ~N[2019-07-15 12:47:38],
      price: %Dinero{amount: 2000, currency: :USD},
      updated_at: ~N[2019-07-15 12:47:38]
    }}
    ```
5. Get from DB:
    ```elixir
    iex> Repo.all(Product)
    [debug] QUERY OK source="products" db=1.2ms queue=1.3ms
    SELECT p0."id", p0."price", p0."inserted_at", p0."updated_at" FROM "products" AS p0 []
    [
      %EctoCustomType.Product{
        __meta__: #Ecto.Schema.Metadata<:loaded, "products">,
        id: 1,
        inserted_at: ~N[2019-07-15 12:47:38],
        price: %Dinero{amount: 2000, currency: :USD},
        updated_at: ~N[2019-07-15 12:47:38]
      }
    ]

    ```

### Dinero.Sigil

```elixir
# Sigils for Dinero
import Dinero.Sigil

iex> ~m[100]UAH
%Dinero{amount: 10000, currency: :UAH}

# Default currency is USD
iex> ~m[100]
%Dinero{amount: 10000, currency: :USD}
```

## Installation

It is [available in Hex](https://hex.pm/packages/dinero) and can be installed
by adding `dinero` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:dinero, "~> 1.3"}
  ]
end
```

Documentation can be found [here](https://hexdocs.pm/dinero)

## Authors
* [Slava Fir][github-fir], MLSDev

## License
Dinero is released under the MIT license. See LICENSE for details.

## About MLSDev

[<img src="https://raw.githubusercontent.com/MLSDev/development-standards/master/mlsdev-logo.png" alt="MLSDev.com">][mlsdev]

[Dinero](https://github.com/MLSDev/dinero) is maintained by MLSDev, Inc. We specialize in providing all-in-one solution in mobile and web development. Our team follows Lean principles and works according to agile methodologies to deliver the best results reducing the budget for development and its timeline. 

Find out more [here][mlsdev] and don't hesitate to [contact us][contact]!

[mlsdev]: https://mlsdev.com
[contact]: https://mlsdev.com/contact-us
[github-fir]: https://github.com/SlavaFir


CHANGELOG.md
# Changelog

### `1.3.2` - 09.09.2019
- `Dinero.to_float/1` added

### `1.3.1` - 09.09.2019
 - `DineroWithCurrency.cast/1` now supports `Dinero` struct
 - support for Jason.Encoder for Ecto embeds

### `1.3.0` - 19.08.2019
 - `Dinero.parse!/2` added
 - `Dinero.parse/2` now conforms to Elixir conventions and returns either `{:ok, %Dinero{}}` or `:error`
 - `DineroWithCurrency.cast/1` now supports `String`, `integer` and `float`
 - parsing Strings like `"1,235,837.50"`
 - default currency for `Dinero.new/2` added (`:USD`)

### `1.2.3` - 18.08.2019
 - `Dinero.equals?/2` and `Dinero.zero?/1` added

### `1.2.2` - 13.08.2019
 - Ability to create `Dinero` struct from String added. See `Dinero.parse/2`
 - Ability to create `Dinero` struct from [scientific notation](https://en.wikipedia.org/wiki/Scientific_notation)

### `1.2.1` - 10.08.2019
 - Convertation fixed
 - Default currency for sigil is USD
 - `String.Chars` implemented and now `Dinero` can be used as an argument for `to_string/1`

### `1.2.0` - 22.07.2019
 - Now it contains 2 methods `Dinero.multiply/2` and `Dinero.multiply/3` for truncations and rounding up results of multiplying
 - Now there are `Dinero.divide/2` and `Dinero.divide/3` for truncations and rounding up results of dividing
 - Small docs fixes

### `1.1.0` - 19.07.2019
- Ability to add/subtract int or float values removed. Now `Dinero.add/2` and `Dinero.subtract/2` use only `Dinero` structs for calculations

### `1.0.0` - 15.07.2019
- Initial Release


test/sigil_test.exs
defmodule Dinero.SigilTest do
  use ExUnit.Case, async: true
  doctest Dinero.Sigil

  import Dinero.Sigil

  test "valid sigil" do
    assert ~m[1_100.256]uah == Dinero.new(1100.25, :UAH)
    assert ~m[1]USD == Dinero.new(1, :USD)
  end

  test "invalid sigil" do
    invalid_sigil = quote(do: ~m[shit]usd)
    assert %ArgumentError{} = catch_error(Code.eval_quoted(invalid_sigil))
  end

  test "default currency" do
    assert ~m[100] == Dinero.new(100, :USD)
  end
end


test/utils_test.exs
defmodule UtilsTest do
  use ExUnit.Case
  alias Dinero.Utils

  test "convert_currency_to_coins for integer" do
    assert Utils.convert_currency_to_coins(12_356) == 1_235_600
  end

  test "convert_currency_to_coins for float" do
    assert Utils.convert_currency_to_coins(2568.7) == 256_870
    assert Utils.convert_currency_to_coins(0.02) == 2
    assert Utils.convert_currency_to_coins(0.0002) == 0
    assert Utils.convert_currency_to_coins(1.0e3) == 100_000
  end
end


test/dinero_test.exs
defmodule DineroTest do
  use ExUnit.Case
  doctest Dinero

  test "new with integer amount" do
    assert Dinero.new(123, :USD) == %Dinero{amount: 12_300, currency: :USD}
  end

  test "new with float amount" do
    assert Dinero.new(0.02, :USD) == %Dinero{amount: 2, currency: :USD}
    assert Dinero.new(2568.7, :USD) == %Dinero{amount: 256_870, currency: :USD}
  end

  test "new with string currency" do
    assert Dinero.new(1235.32, "uah") == %Dinero{amount: 123_532, currency: :UAH}
  end

  test "new with invalid currency" do
    assert_raise ArgumentError, fn ->
      Dinero.new(123, :ABC)
    end
  end

  test "default currency" do
    assert %Dinero{amount: 100_00, currency: :USD} == Dinero.new(100)
  end

  test "add the same currencies" do
    a = Dinero.new(123, :USD)
    b = Dinero.new(321, :USD)
    assert Dinero.add(a, b) == Dinero.new(444, :USD)
  end

  test "add different currencies" do
    a = Dinero.new(1, :USD)
    b = Dinero.new(2, :EUR)

    assert_raise FunctionClauseError, fn ->
      Dinero.add(a, b)
    end
  end

  test "subtract the same currencies" do
    a = Dinero.new(100, :USD)
    b = Dinero.new(20, :USD)
    assert Dinero.subtract(a, b) == Dinero.new(80, :USD)
  end

  test "subtract different currencies" do
    a = Dinero.new(100, :USD)
    b = Dinero.new(20, :EUR)

    assert_raise FunctionClauseError, fn ->
      Dinero.subtract(a, b)
    end
  end

  test "multiply integer" do
    a = Dinero.new(100, :USD)
    b = 5
    assert Dinero.multiply(a, b) == Dinero.new(500, :USD)
  end

  test "multiply round float" do
    d1 = Dinero.new(100, :USD)
    assert Dinero.multiply(d1, 0.05, :round_up) == Dinero.new(5, :USD)
    assert Dinero.multiply(d1, 1.005, :round_up) == Dinero.new(100.5, :USD)
    assert Dinero.multiply(d1, 1.5, :round_up) == Dinero.new(150, :USD)

    d2 = Dinero.new(0.15, :USD)
    assert Dinero.multiply(d2, 1.05, :round_up) == Dinero.new(0.16, :USD)

    assert Dinero.multiply(Dinero.new(8.93, :USD), 121.75, :round_up) == Dinero.new(1087.23, :USD)
    assert Dinero.multiply(Dinero.new(8.93, :USD), 161.75, :round_up) == Dinero.new(1444.43, :USD)

    assert Dinero.multiply(Dinero.new(17.26, :USD), 125.99, :round_up) ==
             Dinero.new(2174.59, :USD)
  end

  test "multiply trunc float" do
    a1 = Dinero.new(100, :USD)
    a2 = Dinero.new(0.15, :USD)
    assert Dinero.multiply(a1, 1.005) == Dinero.new(100.49, :USD)
    assert Dinero.multiply(a2, 1.05) == Dinero.new(0.157, :USD)

    assert Dinero.multiply(Dinero.new(8.93, :USD), 121.75) == Dinero.new(1087.22, :USD)
    assert Dinero.multiply(Dinero.new(8.93, :USD), 161.75) == Dinero.new(1444.42, :USD)
    assert Dinero.multiply(Dinero.new(17.26, :USD), 125.99) == Dinero.new(2174.58, :USD)
  end

  test "divide trunc" do
    hours_per_month = 168

    assert Dinero.divide(Dinero.new(3000, :USD), hours_per_month) == Dinero.new(17.85, :USD)
    assert Dinero.divide(Dinero.new(1500, :USD), hours_per_month) == Dinero.new(8.92, :USD)
    assert Dinero.divide(Dinero.new(2500, :USD), hours_per_month) == Dinero.new(14.88, :USD)
  end

  test "divide round up" do
    hours_per_month = 168

    assert Dinero.divide(Dinero.new(3000, :USD), hours_per_month, :round_up) ==
             Dinero.new(17.86, :USD)

    assert Dinero.divide(Dinero.new(1500, :USD), hours_per_month, :round_up) ==
             Dinero.new(8.93, :USD)

    assert Dinero.divide(Dinero.new(2500, :USD), hours_per_month, :round_up) ==
             Dinero.new(14.88, :USD)
  end

  test "exchange different currencies" do
    source = Dinero.new(100.34, :USD)
    rate = 25.6
    target = :UAH

    assert Dinero.convert(source, target, rate) == Dinero.new(2568.7, :UAH)

    assert Dinero.convert(Dinero.new(924.06, :UAH), :USD, 1.0 / 25.489178) ==
             Dinero.new(36.25, :USD)
  end

  test "the same currencies" do
    source = Dinero.new(100, :USD)
    rate = 123
    target = :USD

    assert_raise ArgumentError, fn ->
      Dinero.convert(source, target, rate)
    end
  end

  test "to_string/1 and String.Chars protocol implementation" do
    d = Dinero.new(100, :USD)
    assert to_string(d) == "100.00"
    assert "#{d}" == "100.00"
  end

  test "parse/2" do
    assert Dinero.parse("123") == {:ok, Dinero.new(123, :USD)}
    assert Dinero.parse("123") == {:ok, Dinero.new(123, :USD)}
    assert Dinero.parse("1,123", :UAH) == {:ok, Dinero.new(1123, :UAH)}
  end

  test "parse/2 invalid string" do
    assert Dinero.parse("invalid string") == :error
    assert Dinero.parse("invalid string") == :error
  end

  test "parse!/2 valid input" do
    assert Dinero.parse!("123") == Dinero.new(123, :USD)
    assert Dinero.parse!("123") == Dinero.new(123, :USD)
    assert Dinero.parse!("1,123,000", :UAH) == Dinero.new(1123000, :UAH)
    assert Dinero.parse!("256_870") == Dinero.new(256_870, :USD)
    assert Dinero.parse!("321.04") == Dinero.new(321.04, :USD)
    assert Dinero.parse!("321.0569234234") == Dinero.new(321.05, :USD)
    assert Dinero.parse!("100.00") == Dinero.new(100, :USD)
    assert Dinero.parse!("1231.123123.12312.123123") == Dinero.new(1231.12, :USD)
  end

  test "parse!/2 invalid input" do
    assert_raise ArgumentError, fn ->
      Dinero.parse!("invalid string")
    end

    assert_raise ArgumentError, fn ->
      Dinero.parse!("invalid.string")
    end
  end

  test "equals?/2" do
    assert Dinero.equals?(Dinero.new(2568.7, :UAH), Dinero.parse!("2568.70", "UAH"))
    refute Dinero.equals?(Dinero.new(2568.7, :USD), Dinero.parse!("2568.70", "UAH"))
  end

  test "zero?/1" do
    assert Dinero.zero?(%Dinero{amount: 0, currency: :USD})
    refute Dinero.zero?(%Dinero{amount: 1, currency: :UAH})
  end
end


mix.exs
defmodule Dinero.MixProject do
  use Mix.Project

  @version "1.3.2"

  def project do
    [
      app: :dinero,
      version: @version,
      elixir: "~> 1.9",
      start_permanent: Mix.env() == :prod,
      description: description(),
      deps: deps(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],
      package: package(),
      source_url: "https://github.com/MLSDev/dinero",
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp description() do
    "Elixir library for working with Money (slang Dinero)"
  end

  defp package() do
    [
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/MLSDev/dinero"}
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:credo, "~> 1.0.0", only: [:dev, :test], runtime: false},
      {:ecto, "~>3.0", optional: true},
      {:excoveralls, "~> 0.10", only: :test},
      {:ex_doc, "~> 0.19", only: :dev, runtime: false}
    ]
  end

  def docs do
    [
      extras: [
        "README.md": [filename: "readme", title: "Readme"],
        "CHANGELOG.md": [filename: "changelog", title: "Changelog"],
      ],
      main: "readme",
      source_ref: "v#{@version}"
    ]
  end
end


lib/sigil.ex
defmodule Dinero.Sigil do
  @moduledoc """
  Provides sigil to create `Dinero` struct
  """

  @doc ~S"""
  Handles the sigil `~m` for Dinero

  ## Usage

      iex> import Dinero.Sigil
      iex> ~m[100.24]uah
      %Dinero{amount: 10024, currency: :UAH}
      iex> ~m[100]
      %Dinero{amount: 10000, currency: :USD}

  """

  defmacro sigil_m({:<<>>, _, [amount]}, [_ | _] = currency) do
    Macro.escape(Dinero.new(to_float(amount), List.to_atom(currency)))
  end

  defmacro sigil_m({:<<>>, _, [amount]}, []) do
    Macro.escape(Dinero.new(to_float(amount), :USD))
  end

  defp to_float(amount) do
    normalized_str = String.replace(amount, "_", "")

    case Float.parse(normalized_str) do
      {num, _} -> num
      :error -> raise(ArgumentError, "#{normalized_str} can't be recognized as float value")
    end
  end
end


lib/utils.ex
defmodule Dinero.Utils do
  @moduledoc false
  # 2568.7 * 100 must be 256870 and not 256869.9999...
  def convert_currency_to_coins(value) when is_float(value) do
    str_value = Float.to_string(value)

    if String.contains?(str_value, "e") do
      value
      |> trunc()
      |> convert_currency_to_coins()
    else
      parse_general_float(str_value)
    end
  end

  def convert_currency_to_coins(value) when is_integer(value) do
    value * 100
  end

  defp parse_general_float(str_value) do
    list = String.split(str_value, ".")

    first =
      list
      |> List.first()
      |> String.to_integer()
      |> Kernel.*(100)

    last = List.last(list)

    if String.length(last) == 1 do
      first + String.to_integer(last) * 10
    else
      last
      |> String.slice(0, 2)
      |> String.to_integer()
      |> Kernel.+(first)
    end
  end
end


lib/currency.ex
defmodule Dinero.Currency do
  @moduledoc """
  Provides functions to work with supported currencies
  """
  @currencies [
    %{name: "United States dollar", code: :USD},
    %{name: "Euro", code: :EUR},
    %{name: "Zloty", code: :PLN},
    %{name: "Hryvnia", code: :UAH}
  ]

  @spec all() :: list
  @doc ~S"""
  Returns all supported currencies

  ## Example:

      iex> Dinero.Currency.all()
      [
        %{code: :USD, name: "United States dollar"},
        %{code: :EUR, name: "Euro"},
        %{code: :PLN, name: "Zloty"},
        %{code: :UAH, name: "Hryvnia"}
      ]

  """

  def all do
    @currencies
  end

  @spec get!(String.t() | atom) :: map | ArgumentError
  @doc ~S"""
  Returns a map with the name and code of the currency or raises ArgumentError if currency is not found

  ## Example:

      iex> Dinero.Currency.get!("USD")
      %{code: :USD, name: "United States dollar"}
      iex> Dinero.Currency.get!("RUR")
      ** (ArgumentError) currency RUR not found

  """
  def get!(currency) when is_atom(currency) do
    currency
    |> Atom.to_string()
    |> get!
  end

  def get!(currency) when is_binary(currency) do
    Enum.find(all(), fn map ->
      map[:code] |> Atom.to_string() |> String.downcase() == String.downcase(currency)
    end) ||
      raise(ArgumentError, "currency #{currency} not found")
  end
end


lib/ecto/money_with_currency.ex
if Code.ensure_compiled?(Ecto.Type) do
  defmodule Dinero.Ecto.MoneyWithCurrency do
    @behaviour Ecto.Type

    def type, do: :money_with_currency

    def dump(%Dinero{amount: amount, currency: currency}) do
      {:ok, {amount, to_string(currency)}}
    end

    def dump(_), do: :error

    def load({amount, currency}) do
      {:ok, Dinero.new(amount / 100, currency)}
    end

    def load(_), do: :error

    def cast({amount, currency})
        when is_integer(amount) and (is_binary(currency) or is_atom(currency)) do
      {:ok, Dinero.new(amount / 100, currency)}
    end

    # def cast(%{"amount" => amount, "currency" => currency}) when is_integer(amount) and (is_binary(currency) or is_atom(currency)) do
    #  {:ok, Dinero.new(amount, currency)}
    # end

    def cast(_), do: :error
  end
end


lib/ecto/dinero_with_currency.ex
if Code.ensure_compiled?(Ecto.Type) do
  defmodule Dinero.Ecto.DineroWithCurrency do
    @moduledoc """
    Provides a custom type for Ecto to save dinero's amount with currency in database

    ## Example

      1. Create custom Postrgres type
      
      ```
      def up do
        execute "CREATE TYPE public.dinero_with_currency AS (amount integer, currency char(3))"
      end
      
      def down do
        execute "DROP TYPE public.dinero_with_currency"
      end
      ```

      2. Add column to the table

      ```
      alter table (:some_table) do
        add :price, :dinero_with_currency
      end
      ```
      3. Add field to the schema
      ```
      schema "some_table" do
        field :price, Dinero.Ecto.DineroWithCurrency
      end
      ```
    """

    @behaviour Ecto.Type

    def type, do: :dinero_with_currency

    def dump(%Dinero{amount: amount, currency: currency}) do
      {:ok, {amount, to_string(currency)}}
    end

    def dump(_), do: :error

    def load({amount, currency}) do
      {:ok, Dinero.new(amount / 100, currency)}
    end

    def load(%{"amount" => amount, "currency" => currency}) do
      {:ok, Dinero.new(amount / 100, currency)}
    end

    def load(_), do: :error

    def cast({amount, currency})
        when is_integer(amount) and (is_binary(currency) or is_atom(currency)) do
      {:ok, Dinero.new(amount / 100, currency)}
    end

    def cast(amount) when is_binary(amount) do
      Dinero.parse(amount)
    end

    def cast(amount) when is_integer(amount) do
      {:ok, Dinero.new(amount)}
    end

    def cast(amount) when is_float(amount) do
      {:ok, Dinero.new(amount)}
    end

    def cast(%Dinero{} = dinero) do
      {:ok, dinero}
    end

    def cast(_), do: :error
  end
end


lib/dinero.ex
defmodule Dinero do
  alias Dinero.Currency
  alias Dinero.Utils

  @moduledoc """
  `Dinero` is a struct that provides methods for working with currencies

  ## Examples

      iex> d1 = Dinero.new(100, :USD)
      %Dinero{amount: 10000, currency: :USD}
      iex> d2 = Dinero.new(200, :USD)
      %Dinero{amount: 20000, currency: :USD}
      iex> Dinero.add(d1, d2)
      %Dinero{amount: 30000, currency: :USD}


  **Note:** `Dinero` uses coins value for calculations. So when you
  create a new `Dinero` struct with 100 USD it automatically transforms this into 10000 cents
  """

  @enforce_keys [:amount, :currency]
  defstruct [:amount, :currency]

  @type t :: %__MODULE__{
          amount: integer,
          currency: atom
        }

  @spec new(integer | float, atom | String.t()) :: t
  @doc ~S"""
  Creates a new `Dinero` struct with provided currency.
  If currency is not supported, ArgumentError will be raised

  ## Examples

      iex> Dinero.new(100, :USD)
      %Dinero{amount: 10000, currency: :USD}
      iex> Dinero.new(1.0e4, :USD) 
      %Dinero{amount: 1000000, currency: :USD}
      iex> Dinero.new(100, :RUR)
      ** (ArgumentError) currency RUR not found
    
  """
  def new(amount, currency \\ :USD)
      when (is_integer(amount) or is_float(amount)) and (is_atom(currency) or is_binary(currency)) do
    %Dinero{
      amount: Utils.convert_currency_to_coins(amount),
      currency: get_currency_code(currency)
    }
  end

  @spec add(t, t) :: t
  @doc ~S"""
  Adds two `Dinero` structs

  ## Examples

      iex> d1 = Dinero.new(100, :USD)
      %Dinero{amount: 10000, currency: :USD}
      iex> d2 = Dinero.new(20, :USD)
      %Dinero{amount: 2000, currency: :USD}
      iex> Dinero.add(d1, d2)
      %Dinero{amount: 12000, currency: :USD}
    
  """
  def add(%Dinero{amount: a, currency: currency}, %Dinero{amount: b, currency: currency}) do
    %Dinero{amount: a + b, currency: get_currency_code(currency)}
  end

  @spec subtract(t, t) :: t
  @doc ~S"""
  Subtracts one `Dinero` from another

  ## Examples

      iex> d1 = Dinero.new(100, :USD)
      %Dinero{amount: 10000, currency: :USD}
      iex> d2 = Dinero.new(20, :USD)
      %Dinero{amount: 2000, currency: :USD}
      iex> Dinero.subtract(d1, d2)
      %Dinero{amount: 8000, currency: :USD}

  """
  def subtract(%Dinero{amount: a, currency: currency}, %Dinero{amount: b, currency: currency}) do
    %Dinero{amount: a - b, currency: get_currency_code(currency)}
  end

  @spec multiply(t, integer | float) :: t
  @doc ~S"""
  Multiplies a `Dinero` by a value and truncates the result

  ## Examples

      iex> d = Dinero.new(120, :USD)
      %Dinero{amount: 12000, currency: :USD}
      iex> Dinero.multiply(d, 4)
      %Dinero{amount: 48000, currency: :USD}
      iex> d = Dinero.new(100, :USD)
      %Dinero{amount: 10000, currency: :USD}
      iex> Dinero.multiply(d, 1.005)
      %Dinero{amount: 10049, currency: :USD}

  """
  def multiply(%Dinero{amount: a, currency: currency}, value)
      when is_integer(value) or is_float(value) do
    %Dinero{amount: trunc(a * value), currency: get_currency_code(currency)}
  end

  @spec multiply(t, integer | float, :round_up) :: t
  @doc ~S"""
  Multiplies a `Dinero` by a value and rounds up the result

  ## Examples

      iex> d = Dinero.new(100, :USD)
      %Dinero{amount: 10000, currency: :USD}
      iex> Dinero.multiply(d, 1.005, :round_up)
      %Dinero{amount: 10050, currency: :USD}

  """
  def multiply(%Dinero{amount: a, currency: currency}, value, :round_up)
      when is_number(value) do
    %Dinero{amount: round(a * value), currency: get_currency_code(currency)}
  end

  @spec divide(t, integer | float) :: t
  @doc ~S"""
  Divides `Dinero` by a value and truncates the result

  ## Examples

      iex> d = Dinero.new(100.24, :USD)
      %Dinero{amount: 10024, currency: :USD}
      iex> Dinero.divide(d, 3)
      %Dinero{amount: 3341, currency: :USD}
      iex> Dinero.divide(d, 5)
      %Dinero{amount: 2004, currency: :USD}

  """
  def divide(%Dinero{amount: a, currency: currency}, value)
      when is_integer(value) or is_float(value) do
    %Dinero{amount: trunc(a / value), currency: get_currency_code(currency)}
  end

  @spec divide(t, integer | float, atom) :: t
  @doc ~S"""
  Divides `Dinero` by a value and rounds up the result

  ## Examples

      iex> d = Dinero.new(100.24, :USD)
      %Dinero{amount: 10024, currency: :USD}
      iex> Dinero.divide(d, 3, :round_up)
      %Dinero{amount: 3341, currency: :USD}
      iex> Dinero.divide(d, 5, :round_up)
      %Dinero{amount: 2005, currency: :USD}

  """
  def divide(%Dinero{amount: a, currency: currency}, value, round_up)
      when (is_integer(value) or is_float(value)) and is_atom(round_up) do
    %Dinero{amount: round(a / value), currency: get_currency_code(currency)}
  end

  @spec convert(t, atom | String.t(), float) :: t
  @doc ~S"""
  Converts value of `Dinero` to target currency using exchange_rate and truncates the result

  ## Examples

      iex> d = Dinero.new(100, :USD)
      %Dinero{amount: 10000, currency: :USD}
      iex> Dinero.convert(d, :UAH, 26.2)
      %Dinero{amount: 262000, currency: :UAH}
      iex> Dinero.convert(d, :USD, 26.2)
      ** (ArgumentError) target currency must be different from source currency

  """
  def convert(%Dinero{} = d, target, exchange_rate) do
    if get_currency_code(d.currency) != get_currency_code(target) do
      %Dinero{
        amount: trunc(d.amount * exchange_rate),
        currency: get_currency_code(target)
      }
    else
      raise(ArgumentError, "target currency must be different from source currency")
    end
  end

  @spec parse!(String.t(), atom) :: t
  @doc ~S"""
  Creates `Dinero` from String that represents integer or float. If a string can't be parsed ArgumentError is raised
  If the second param is not provided it uses USD as default currency

  ## Examples

      iex> Dinero.parse!("123.23")
      %Dinero{amount: 12323, currency: :USD}
      iex> Dinero.parse!("112")
      %Dinero{amount: 11200, currency: :USD}
      iex> Dinero.parse!("2", :UAH)
      %Dinero{amount: 200, currency: :UAH}
      iex> Dinero.parse!("100.00")  
      %Dinero{amount: 10000, currency: :USD}
      iex> Dinero.parse!("invalid string")
      ** (ArgumentError) invalid string. it must contain string that represents integer or float

  """
  def parse!(amount, currency \\ :USD) when is_binary(amount) do
    case parse(amount, currency) do
      {:ok, %Dinero{} = dinero} ->
        dinero

      :error ->
        raise(
          ArgumentError,
          "invalid string. it must contain string that represents integer or float"
        )
    end
  end

  @spec parse(String.t(), atom) :: {:ok, t} | :error
  @doc """
  Same as `parse!/2`, but returns either `{:ok, %Dinero{}}` or `:error`

  ## Examples

      iex> Dinero.parse("123.23")
      {:ok, %Dinero{amount: 12323, currency: :USD}}
      iex> Dinero.parse("invalid string")
      :error

  """
  def parse(amount, currency \\ :USD) when is_binary(amount) do
    amount = String.replace(amount, ~r/[_,]/, "")

    amount =
      if String.contains?(amount, ".") do
        Float.parse(amount)
      else
        Integer.parse(amount)
      end

    case amount do
      {a, _} -> {:ok, Dinero.new(a, currency)}
      :error -> :error
    end
  end

  @spec equals?(t, t) :: boolean
  @doc """
  Compares two Dinero structs.

  Returns `true`, if amounts and currencies match. Otherwise returns `false`.
  """
  def equals?(%Dinero{} = a, %Dinero{} = b) do
    a.amount == b.amount && get_currency_code(a.currency) == get_currency_code(b.currency)
  end

  @spec zero?(t) :: boolean
  @doc """
  Returns `true`, if amount is zero.
  """
  def zero?(%Dinero{} = a), do: a.amount == 0

  defp get_currency_code(currency) do
    Currency.get!(currency).code
  end

  @spec to_float(t) :: float
  @doc """
  Returns float representation of Dinero.

  ## Examples

      iex> import Dinero.Sigil
      iex> Dinero.to_float(~m[100.14]USD)
      100.14
      iex> Dinero.to_float(~m[0.01]uah)
      0.01

  """
  def to_float(%Dinero{amount: amount}) do
    amount / 100.0
  end
end

defimpl String.Chars, for: Dinero do
  def to_string(dinero) do
    :erlang.float_to_binary(dinero.amount / 100, decimals: 2)
  end
end

if Code.ensure_compiled?(Jason.Encoder) do
  require Protocol
  Protocol.derive(Jason.Encoder, Dinero, only: [:amount, :currency])
end
